### PR TITLE
AR-1300: Add a keyword search over text fields in the prisoner index

### PR DESCRIPTION
**Pull request by timharrison-moj**
* Adds a keyword search endpoint, service and tests.
* Filters results by a mandatory set of prison identifiers.
* Implements a pageable result set.
* One of two new endpoints & services required initially for IMS, but available to other clients.
* Requires role GLOBAL_SEARCH.
* Marked as WIP in the swagger docs - minor changes may follow after testing period.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+      andWords.takeIf { !it.isNullOrBlank() }?.let {
+        // Will match only if all of these words are present across all text fields
+        keywordQuery.must(
+          QueryBuilders.boolQuery()
+            .minimumShouldMatch(1)
+            .should(
+              QueryBuilders.multiMatchQuery(it)
+                .type(MultiMatchQueryBuilder.Type.CROSS_FIELDS)
+                .lenient(true)
+                .fuzzyTranspositions(fuzzyMatch!!)
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt:111]: minor, I guess this isn't nullable so no need for !! ?

**Review by andymarke — APPROVED** [review]
LGTM

Excellent feature!

I think I would spend a bit more time on the tests doing 2 things:

1) Dynamically loading data to fit your scenarios
2) Don't assert against resources files, just assert you have say "2 matching prisoners" with "prisoner numbers matching". Since for the feature we really don't care too much about each response JSON, but we really do care that an AND works correctly with say a mixed case PNC etc etc 

Take a look at probation-offender-search - I think those tests are much better model than the prisoner searches which IMHO are way to static. In particular look at the equivalent OffenderSearchPhraseAPIIntegrationTest test

File: src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt (hunk 1 above)
Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt:111]: Think I altered this to default if null in the last round of testing, so will remove the !!.

Comment (timharrison-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/prisonersearch/services/KeywordService.kt]: I checked this again and it is a nullable type `Boolean?` - which may not be provided by the client so I've defaulted it to false if it is ever null.

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**